Trigger cold reboots upon changes in KVM environment parameters

sysvinit opened the pull request:
This change adds support to fc-agent for a new mechanism in fc.qemu which signals selected boot-time parameters from the KVM environment, similar to the existing qemu binary generation handling. The corresponding side in fc.qemu will seed the parameters once at boot and then periodically write them into /run at runtime through the qemu guest agent. If fc-agent detects changes, then it will schedule a cold reboot in a maintenance window so the KVM host can carry out changes which require a restart of the qemu process.

PL-131857

@flyingcircusio/release-managers

## Release process

Impact:

Changelog:
- VMs will now automatically schedule reboots during maintenance when changes in KVM parameters such as CPU flags or VM backing storage (i.e. SSD or HDD) are detected (PL-131857).

### PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [x] ticket state set to *Pull request ready*
- [x] if ticket is more urgent than within the next few days, directly contact a member of the Platform team

## Design notes

- [x] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [x] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - This is an extension of the existing qemu binary generation upgrade mechanism, where the hypervisor signals the state of the guest-external KVM environment through the qemu guest agent.
- [x] Security requirements tested? (EVIDENCE)
  - Additional unit tests, in particular to test compatibility with the previous qemu binary generation handling mechanism, and the upgrade path from the existing code to the new code in fc.qemu.
  - Verified manually in DEV.

ctheune commented on pkgs/fc/agent/fc/maintenance/maintenance.py:
```diff
@@ -23,6 +24,10 @@
 
+# If the value of these parameters in the ENC data change, then a cold
+# reboot should be scheduled
+ENC_PARAMETERS_FOR_COLD_REBOOT = ["cpu_model", "rbd_pool"]
```
As discussed: maintaining the parameters on the agent running within the VM is prone to us not being able to manage infrastructure aspects sufficiently independent from the guest agent. Meaning: if we decide that further parameters need to be able to trigger cold reboots, then we need to release updates to older platforms. Please let the inner agent just use whatever data is passed into the guest-properties file to make a decision for a cold boot.

sysvinit replied:
Fixed in [25e49b3](https://github.com/flyingcircusio/fc-nixos/pull/1124/commits/25e49b35f3b49b6e605f9fe2fcd7e8e3cf375a44)

ctheune commented on pkgs/fc/agent/fc/maintenance/maintenance.py:
```diff
@@ -56,53 +61,107 @@ def request_reboot_for_cpu(log, enc) -> Optional[Request]:
+    if booted_properties:
+        current_properties = enc["parameters"]
+        changed_properties = []
+        for key in ENC_PARAMETERS_FOR_COLD_REBOOT:
```
- newly added keys should likely trigger a reboot
- keys that go away maybe should not trigger a reboot?
- when making changes from the earlier comment this needs to exclude the binary generation OR decide that binary generation can also be downgraded and make the old/new style paths more separate (and then we can retire the old path in 24.11 or so)
- i think it's fine if we schedule gratuitous reboots to introduce this feature

sysvinit replied:
Addressed in [25e49b3](https://github.com/flyingcircusio/fc-nixos/pull/1124/commits/25e49b35f3b49b6e605f9fe2fcd7e8e3cf375a44).

sysvinit commented:
After further discussion we've decided to revise some of the logic in here, and instead of using ENC as the source of the desired state of the world, we should use an extra properties file fed into the VM by the qemu guest agent from the KVM layer. This means we're no longer limited to only comparing values inside ENC, and means we can detect new guest property values being added by the hypervisor at runtime which were not present at boot time.

sysvinit committed:
agent: trigger cold reboots when KVM environment parameters change

Add support for a mechanism in fc.qemu similar to the existing qemu
binary generation handling which additionally seeds various parameters
related to the KVM environment at boot time and which writes these
parameters into /run periodically at runtime. Request a cold reboot
when the KVM environment parameters change at runtime to allow the
hypervisor to perform changes which require the qemu process to be
offline or restarted.

PL-131857

sysvinit committed:
agent: tests for KVM environment parameter change handling

PL-131857
